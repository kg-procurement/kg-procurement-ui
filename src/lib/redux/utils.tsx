import { SerializedError } from '@reduxjs/toolkit'
import { FetchBaseQueryError } from '@reduxjs/toolkit/query'

export type ZodParseError = Extract<
  FetchBaseQueryError,
  {
    status: 'PARSING_ERROR'
    originalStatus: number
    data: string
    error: string
  }
>

export function isZodParseError(
  err: SerializedError | FetchBaseQueryError | undefined,
): err is ZodParseError {
  if (!err) return false
  return (
    'data' in err &&
    'error' in err &&
    'originalStatus' in err &&
    'status' in err
  )
}

export function extractZodError(err: ZodParseError): React.ReactNode {
  const errorList = err.error.split('\n').slice(0, -1)
  return (
    <div className="flex flex-col">
      <div className="font-semibold">
        Response does not match validator schema:
      </div>
      <ul className="list-inside list-disc">
        {errorList.map((e, i) => (
          <li key={i}>{e}</li>
        ))}
      </ul>
    </div>
  )
}

export function isErrorWithMessage(
  err: FetchBaseQueryError,
): err is { status: number, data: { error: string } } {
  if (!err.data) return false
  return (
    typeof err.data === 'object' &&
    'error' in err.data &&
    typeof err.data.error === 'string'
  )
}
