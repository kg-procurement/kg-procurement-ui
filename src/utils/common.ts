export const noop = () => {}

export const formatPrice = (
  amount: number,
  quantityPrice: number,
  currencyCode: string,
  uomName: string
): string => {
  const formattedPrice = new Intl.NumberFormat(
    currencyCode === 'IDR' ? 'id-ID' : 'en-US',
    {
      style: 'currency',
      currency: currencyCode,
    }
  ).format(amount);

  return quantityPrice > 1
    ? `${formattedPrice} / ${quantityPrice} ${uomName}`
    : `${formattedPrice} / ${uomName}`;
};
