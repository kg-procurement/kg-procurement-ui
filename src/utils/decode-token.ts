export interface Claims {
  sub: string
  exp: number
}

export function decodeToken(token: string): Claims | null {
  try {
    const payload = JSON.parse(atob(token.split('.')[1]))
    return {
      sub: payload.sub,
      exp: payload.exp,
    }
  }
  catch (error) {
    console.error('Failed to decode token:', error)
    throw error
  }
}
