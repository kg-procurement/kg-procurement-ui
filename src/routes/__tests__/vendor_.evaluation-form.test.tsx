import { render, screen, within } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { describe, expect, it } from 'vitest'

import { withWrappers } from '@/lib/testing/utils.tsx'
import { waitForNoLoadingOverlay } from '@/lib/testing/wait-for.ts'

import EvaluationFormPage from '../vendor_.evaluation-form.tsx'

describe('<EvaluationFormPage/>', () => {
  it('should render the header section with the logo', () => {
    render(withWrappers(<EvaluationFormPage />))

    const logo = screen.getByAltText('Kompas Gramedia Logo Background')
    expect(logo).toBeInTheDocument()
    expect(logo).toHaveAttribute('src', '/kompas-gramedia-logo-bg.svg')

    expect(screen.getByText(/Vendor Performance\s*Evaluation Form/i)).toBeInTheDocument()
    expect(screen.getByText('Vendor A')).toBeInTheDocument()
  })

  it('should render the form penilaian table content properly', async () => {
    const { container } = render(
      withWrappers(<EvaluationFormPage />, { withRoot: true }),
    )
    await waitForNoLoadingOverlay()
    expect(container.innerText.trim()).toMatchSnapshot()
  })

  it('should render the footer', () => {
    render(withWrappers(<EvaluationFormPage />))
    expect(screen.getByText(/© 2024 KOMPAS/i)).toBeInTheDocument()
  })

  it('should render handle update value when other radio button clicked', async () => {
    render(withWrappers(<EvaluationFormPage />, { withRoot: true }))

    await waitForNoLoadingOverlay()

    // const rows = screen.getAllByRole('row')

    screen.debug(undefined, 100000)

    const radioGroups = screen.getAllByRole('radiogroup')
    const firstRadioButton = within(radioGroups[0]).getByRole('radio')
    const secondRadioButton = within(radioGroups[1]).getByRole('radio')
    // const secondRadioGroup = within(rows[1]).getAllByRole('radiogroup')

    await userEvent.click(firstRadioButton)
    await userEvent.click(secondRadioButton)
  })

  it('should render the footer', () => {
    render(withWrappers(<EvaluationFormPage />))
    expect(screen.getByText(/© 2024 KOMPAS/i)).toBeInTheDocument()
  })
})
