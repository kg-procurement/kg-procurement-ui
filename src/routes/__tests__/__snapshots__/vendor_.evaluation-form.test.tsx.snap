// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EvaluationFormPage/> > should render the form penilaian table content properly 1`] = `
"Email Status
Products
Vendors
Evaluation Form
Vendor PerformanceEvaluation Form
Vendor A
Kriterian penilaian1234Kesesuaian produk/jasa dengan spesifikasiKualitas produk/jasaKetepatan waktu pengiriman/penyelesaianKompetitifitas hargaResponsivitas dan kemampuan komunikasiKemampuan dalam menangani masalah/keluhanKelengkapan Barang (Vendor Konvensional)Harga (Vendor Konvensional & Online)Term of Payment (Vendor Konvensional)Reputasi (Vendor Online)Ketersediaan Barang (Vendor Konvensional)Kualitas layanan after services* (jika ada?)
Â© 2024 KOMPAS
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
Contact us (Email):"
`;
